src/Strategy/Googlesource/RepoManifest.hs

```diff
@@ -0,0 +1,218 @@
+    }
+
+-- DTD for the Repo manifest.xml file: https://gerrit.googlesource.com/git-repo/+/master/docs/manifest-format.md
+-- Note that the DTD is only "roughly adhered to" according to the documentation. For example, the DTD says that
+-- there will be zero or more project and remote tags (it uses a `*`), but the documentation specifies at least one
+-- for both of these tags (which should be denoted by a `+` in the DTD).
+data RepoManifest = RepoManifest
+  { manifestDefault  :: Maybe ManifestDefault
+  , manifestRemotes  :: [ManifestRemote]
+  , manifestProjects :: [ManifestProject]
+  , manifestIncludes :: [ManifestInclude]
+  } deriving (Eq, Ord, Show, Generic)
+
+data ManifestRemote = ManifestRemote
+  { remoteName     :: Text
+  , remoteFetch    :: Text
+  , remoteRevision :: Maybe Text
+  } deriving (Eq, Ord, Show, Generic)
+
+data ManifestDefault = ManifestDefault
+  { defaultRemote   :: Maybe Text
+  , defaultRevision :: Maybe Text
+  } deriving (Eq, Ord, Show, Generic)
+
+data ManifestProject = ManifestProject
+  { projectName     :: Text
+  , projectPath     :: Maybe Text
+  , projectRemote   :: Maybe Text
+  , projectRevision :: Maybe Text
+  } deriving (Eq, Ord, Show, Generic)
+
+data ManifestInclude = ManifestInclude { includeName :: Text } deriving (Eq, Ord, Show, Generic)
+
+data ValidatedProject = ValidatedProject
+  { validatedProjectName     :: Text
+  , validatedProjectPath     :: Text
+  , validatedProjectUrl      :: Text
+  , validatedProjectRevision :: Text
+  } deriving (Eq, Ord, Show, Generic)
+
+-- If a project does not have a path, then use its name for the path
+projectPathOrName :: ManifestProject -> Text
+projectPathOrName (ManifestProject { projectPath = Nothing, projectName = name }) = name
+projectPathOrName (ManifestProject { projectPath = Just path }) = path
+
+-- A project's revision comes from the first of these that we encounter:
+--   * If the project has a revision attribute, then use that
+--   * If the project has a remote attribute and the remote it points to has a revision attribute, use that
+--   * If the project does not have a remote attribute and the default remote has a revision attribute, use that
+--   * Otherwise, use the revision attribute on the project's default tag
+--   That leaves these error cases:
+--   * If the project does not have a remote attribute and there is no default remote, then blow up
+--   * If the project does not have a revision and there is no default revision from either its remote or the default, then blow up
+revisionForProject :: RepoManifest -> ManifestProject -> Maybe Text
+revisionForProject manifest project =
+      projectRevision project
+  <|> (remoteForProject manifest project >>= remoteRevision)
+  <|> (manifestDefault manifest >>= defaultRevision)
+
+-- The URL for a project is the project's name appended to the fetch attribute of the project's remote
+urlForProject :: RepoManifest -> ManifestProject -> Maybe Text
+urlForProject manifest project =
```

Comment:
[style] another way to write this (with `Maybe` do notation)
```hs
urlForProject manifest project = do
  remote <- remoteForProject manifest project
  pure $ remoteFetch r <> "/" <> projectName project
```

Reply:
Again, so much nicer :)